pyface/ui/qt4/data_view/data_view_widget.py

```diff
@@ -45,10 +45,56 @@
 }
 
 
+class DataViewTreeView(QTreeView):
+    """ QTreeView subclass that handles drag and drop via DropHandlers. """
+
+    _widget = None
+
+    def dragEnterEvent(self, event):
+        if self._widget is not None:
+            widget = self._widget
+            for drop_handler in widget.drop_handlers:
+                if drop_handler.can_handle_drop(event, self):
+                    event.acceptProposedAction()
+                    return
+        super().dragEnterEvent(event)
+
+    def dragMoveEvent(self, event):
+        if self._widget is not None:
+            widget = self._widget
+            for drop_handler in widget.drop_handlers:
+                if drop_handler.can_handle_drop(event, self):
+                    event.acceptProposedAction()
+                    return
+        super().dragMoveEvent(event)
+
+    def dragLeaveEvent(self, event):
+        if self._widget is not None:
+            widget = self._widget
+            for drop_handler in widget.drop_handlers:
+                if drop_handler.can_handle_drop(event, self):
+                    event.acceptProposedAction()
+                    return
+        super().dragLeaveEvent(event)
+
+    def dropEvent(self, event):
+        if self._widget is not None:
+            widget = self._widget
+            for drop_handler in widget.drop_handlers:
+                if drop_handler.can_handle_drop(event, self):
+                    drop_handler.handle_drop(event, self)
+                    event.acceptProposedAction()
+                    return
+        super().dropEvent(event)
+
+
 @provides(IDataViewWidget)
 class DataViewWidget(MDataViewWidget, Widget):
     """ The Qt implementation of the DataViewWidget. """
 
+    #: Factory for the underlying Qt control, to facilitate replacement
```

Comment:
I'm not sure how this is facilitating replacement. Isn't this facilitating handling of the drop events?

Reply:
One of the lessons learned from the TabularEditor, etc. was that you sometimes want to replace the Qt `QTreeView` control with something different but still adhering to the `QAbstractItemView` API; without a hook for a factory for this control, we need to subclass everything and re-implement `_create_control`; by adding a hook for a factory you just need to set a non-default value for the factory.

We were probably going to need this anyway; the subclassing of the `QTreeView` just meant that this was a good time to add the functionality.